File: ChatClientVB/Services/ChatService.vb
Imports System.Net
Imports ChatClientVB
Imports Microsoft.AspNet.SignalR.Client

Public Class ChatService
    Implements IChatService

    Public Event ParticipantDisconnected(name As String) Implements IChatService.ParticipantDisconnected
    Public Event ParticipantLoggedIn(participant As User) Implements IChatService.ParticipantLoggedIn
    Public Event ParticipantLoggedOut(name As String) Implements IChatService.ParticipantLoggedOut
    Public Event ParticipantReconnected(name As String) Implements IChatService.ParticipantReconnected
    Public Event ConnectionReconnecting() Implements IChatService.ConnectionReconnecting
    Public Event ConnectionReconnected() Implements IChatService.ConnectionReconnected
    Public Event ConnectionClosed() Implements IChatService.ConnectionClosed
    Public Event NewTextMessage(sender As String, msg As String, mt As MessageType) Implements IChatService.NewTextMessage
    Public Event NewImageMessage(sender As String, img As Byte(), mt As MessageType) Implements IChatService.NewImageMessage
    Public Event ParticipantTyping(name As String) Implements IChatService.ParticipantTyping

    Private hubProxy As IHubProxy
    Private connection As HubConnection
    Private url As String = "http://localhost:8080/signalchat"

    Public Async Function ConnectAsync() As Task Implements IChatService.ConnectAsync
        connection = New HubConnection(url)
        hubProxy = connection.CreateHubProxy("ChatHub")
        hubProxy.On(Of User)("ParticipantLogin", Sub(u) RaiseEvent ParticipantLoggedIn(u))
        hubProxy.On(Of String)("ParticipantLogout", Sub(n) RaiseEvent ParticipantLoggedOut(n))
        hubProxy.On(Of String)("ParticipantDisconnection", Sub(n) RaiseEvent ParticipantDisconnected(n))
        hubProxy.On(Of String)("ParticipantReconnection", Sub(n) RaiseEvent ParticipantReconnected(n))
        hubProxy.On(Of String, String)("BroadcastTextMessage",
                                       Sub(n, m) RaiseEvent NewTextMessage(n, m, MessageType.Broadcast))
        hubProxy.On(Of String, Byte())("BroadcastPictureMessage",
                                       Sub(n, m) RaiseEvent NewImageMessage(n, m, MessageType.Broadcast))
        hubProxy.On(Of String, String)("UnicastTextMessage",
                                       Sub(n, m) RaiseEvent NewTextMessage(n, m, MessageType.Unicast))
        hubProxy.On(Of String, Byte())("UnicastPictureMessage",
                                       Sub(n, m) RaiseEvent NewImageMessage(n, m, MessageType.Unicast))
        hubProxy.On(Of String)("ParticipantTyping", Sub(p) RaiseEvent ParticipantTyping(p))

        AddHandler connection.Reconnecting, AddressOf Reconnecting
        AddHandler connection.Reconnected, AddressOf Reconnected
        AddHandler connection.Closed, AddressOf Disconnected

        ServicePointManager.DefaultConnectionLimit = 10
        Await connection.Start()
    End Function

    Private Sub Disconnected()
        RaiseEvent ConnectionClosed()
    End Sub

    Private Sub Reconnecting()
        RaiseEvent ConnectionReconnecting()
    End Sub

    Private Sub Reconnected()
        RaiseEvent ConnectionReconnected()
    End Sub

    Public Async Function LoginAsync(name As String, photo As Byte()) As Task(Of List(Of User)) Implements IChatService.LoginAsync
        Dim users = Await hubProxy.Invoke(Of List(Of User))("Login", New Object() {name, photo})
        Return users
    End Function

    Public Async Function LogoutAsync() As Task Implements IChatService.LogoutAsync
        Await hubProxy.Invoke("Logout")
    End Function

    Public Async Function SendBroadcastMessageAsync(msg As String) As Task Implements IChatService.SendBroadcastMessageAsync
        Await hubProxy.Invoke("BroadcastTextMessage", msg)
    End Function

    Public Async Function SendBroadcastMessageAsync(img As Byte()) As Task Implements IChatService.SendBroadcastMessageAsync
        Await hubProxy.Invoke("BroadcastImageMessage", img)
    End Function

    Public Async Function SendUnicastMessageAsync(recepient As String, msg As String) As Task Implements IChatService.SendUnicastMessageAsync
        Await hubProxy.Invoke("UnicastTextMessage", New Object() {recepient, msg})
    End Function

    Public Async Function SendUnicastMessageAsync(recepient As String, img As Byte()) As Task Implements IChatService.SendUnicastMessageAsync
        Await hubProxy.Invoke("UnicastImageMessage", New Object() {recepient, img})
    End Function

    Public Async Function TypingAsync(recepient As String) As Task Implements IChatService.TypingAsync
        Await hubProxy.Invoke("Typing", recepient)
    End Function
End Class

File: ChatClientCS/Services/ChatService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using ChatClientCS.Enums;
using ChatClientCS.Models;
using System.Net;
using Microsoft.AspNet.SignalR.Client;

namespace ChatClientCS.Services
{
    public class ChatService : IChatService
    {
        public event Action<string, string, MessageType> NewTextMessage;
        public event Action<string, byte[], MessageType> NewImageMessage;
        public event Action<string> ParticipantDisconnected;
        public event Action<User> ParticipantLoggedIn;
        public event Action<string> ParticipantLoggedOut;
        public event Action<string> ParticipantReconnected;
        public event Action ConnectionReconnecting;
        public event Action ConnectionReconnected;
        public event Action ConnectionClosed;
        public event Action<string> ParticipantTyping;

        private IHubProxy hubProxy;
        private HubConnection connection;
        private string url = "http://localhost:8080/signalchat";

        public async Task ConnectAsync()
        {
            connection = new HubConnection(url);
            hubProxy = connection.CreateHubProxy("ChatHub");
            hubProxy.On<User>("ParticipantLogin", (u) => ParticipantLoggedIn?.Invoke(u));
            hubProxy.On<string>("ParticipantLogout", (n) => ParticipantLoggedOut?.Invoke(n));
            hubProxy.On<string>("ParticipantDisconnection", (n) => ParticipantDisconnected?.Invoke(n));
            hubProxy.On<string>("ParticipantReconnection", (n) => ParticipantReconnected?.Invoke(n));
            hubProxy.On<string, string>("BroadcastTextMessage", (n, m) => NewTextMessage?.Invoke(n, m, MessageType.Broadcast));
            hubProxy.On<string, byte[]>("BroadcastPictureMessage", (n, m) => NewImageMessage?.Invoke(n, m, MessageType.Broadcast));
            hubProxy.On<string, string>("UnicastTextMessage", (n, m) => NewTextMessage?.Invoke(n, m, MessageType.Unicast));
            hubProxy.On<string, byte[]>("UnicastPictureMessage", (n, m) => NewImageMessage?.Invoke(n, m, MessageType.Unicast));
            hubProxy.On<string>("ParticipantTyping", (p) => ParticipantTyping?.Invoke(p));

            connection.Reconnecting += Reconnecting;
            connection.Reconnected += Reconnected;
            connection.Closed += Disconnected;

            ServicePointManager.DefaultConnectionLimit = 10;
            await connection.Start();
        }

        private void Disconnected()
        {
            ConnectionClosed?.Invoke();
        }

        private void Reconnected()
        {
            ConnectionReconnected?.Invoke();
        }

        private void Reconnecting()
        {
            ConnectionReconnecting?.Invoke();
        }

        public async Task<List<User>> LoginAsync(string name, byte[] photo)
        {
            return await hubProxy.Invoke<List<User>>("Login", new object[] { name, photo });
        }

        public async Task LogoutAsync()
        {
            await hubProxy.Invoke("Logout");
        }

        public async Task SendBroadcastMessageAsync(string msg)
        {
            await hubProxy.Invoke("BroadcastTextMessage", msg);
        }

        public async Task SendBroadcastMessageAsync(byte[] img)
        {
            await hubProxy.Invoke("BroadcastImageMessage", img);
        }

        public async Task SendUnicastMessageAsync(string recepient, string msg)
        {
            await hubProxy.Invoke("UnicastTextMessage", new object[] { recepient, msg });
        }

        public async Task SendUnicastMessageAsync(string recepient, byte[] img)
        {
            await hubProxy.Invoke("UnicastImageMessage", new object[] { recepient, img });
        }

        public async Task TypingAsync(string recepient)
        {
            await hubProxy.Invoke("Typing", recepient);
        }
    }
}